packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -196,110 +208,265 @@ class CupertinoAlertDialog extends StatelessWidget {
       children.add(const Padding(padding: const EdgeInsets.only(top: 8.0)));
     }
 
+    return new Container(
+      key: const Key('cupertino_alert_dialog_content_section'),
```

Comment:
I'd say the intent of the test is to guarantee a user experience. If this thing here exists to be around all of the dialog's content and provide a background color behind content widgets, that's what you should look for when you want to check that it has a particular size etc. An ID doesn't translate to user experience.

Reply:
I disagree.  The question of the sizing and positioning of the content area of a dialog has nothing to do with the color.  The coincidental fact that the outer most widget of that area applies a color is pure happenstance - and the test that uses this key is not verifying color. 

More broadly, "user experience" is a very generic term and I don't think it will help us develop a litmus test for an effective philosophy.  A test exists to verify a very specific thing, whether it's user related or architecture related or something else entirely.  

If I were verifying color then it would make sense that I find the widget that applies the color and expect the right color.  If I were verifying a button size then it would make sense that I find the button and expect the right size.  In the case of the content section, I'm verifying the size of the content section, so I should find the content section and expect the right size.  I should not find some lower level widget that happens to currently be the outermost widget of the content section's implementation.  In that case we're establishing a fragility that doesn't benefit the test.  We honestly don't care what the outer widget is.  I mean the test fundamentally doesn't care.  So I don't think we should base the success/failure of the test on a specific outer Widget or color.  If a developer comes along later and wraps this Container with another widget then they'd have to go update the test even though absolutely nothing related to the test has changed.

Reply:
Let's chat in person. I think we're re-iterating the same points.

Reply:
I'm not 100% sure what the question is here, but the answer is definitely not a hard-coded string-based key.

Reply:
Based on discussion between myself, @xster, and @Hixie, we opted not to use Keys. Even though Keys can demarcate a logical area of the UI, they can only do it when that area of the UI is represented by exactly 1 Widget.  It is possible that, in general, 2+ Widgets taken collectively represent an area of the UI. An approach with Keys cannot reflect that scenario.

The legitimate fix to this particular case is to create golden tests.  But for the time-being I will make the tests dependent on the knowledge of root Widgets for each logical area of the UI.